### DISCUSSION TITLE
camu.pitchestimate

### DISCUSSION THREAD

**Discussion by christophertrapani** [Ircam-RnD/catart-mubu#90] [Bug, Issues, and Suggestions]
Hi guys: I've been working on a bit of a workaround solution for a pitch estimation. Once markers have been assigned (using MEL segmentation), yin~ data for each segment is scanned and filtered: first with an Energy threshold (removing silences), second with a Periodicity threshold to remove noisier frequency estimates. A median pitch estimate is taken and all data that passes through the threshold is averaged to get a "pitch estimate" (in float MIDI values) which is then added to the MuBu track as a constant. From this point, markers can be selected using this "pitch estimate" as a target, and the results give a far more consistent pitch impression than FrequencyMean.

Short demo on Forum post: https://discussion.forum.ircam.fr/t/segmentation-frequency-estimation/49270

It's a bit of a slow and heavy process, so any suggestions for streamlining it could be helpful. I've also noticed that segmentation is very fickle: there don't seem to be "good" default values for windowsize and filters; I wonder if there might be some way of automating a process that would scan a file for peaks and mean spectral content on import and set these automatically?

I've made a camu.pitchestimate abstraction and a demo tutorial that I will upload to the yinseg brach; feel free to move them if there's a better place, and thanks — C

**Comment by diemoschwarz** [comment]
great use case and cool video!  I'll have a look when I'll be back from NIME, Mexico after the 10th of June.

**Comment by christophertrapani** [comment]
Hi Diemo — Just curious to know if you've had a chance to take a look at
this pitchestimate... Meanwhile, I am experimenting with segmentation in
bach / ears using chordino (http://www.isophonics.net/nnls-chroma) for
estimates of chord changes and finding it really useful, wondering if a
similar process can be used to integrate chordino data as a Mubu column (or
two, one for C/D/E and one for maj/min/dim)...

Thanks and more soon — C

**Comment by diemoschwarz** [comment]
Hi Chris, that's a well tought-out algorithm!  I'm confident this can be formulated as a single pipo graph. Note that there is already a yin.periodicitythreshold. I found that your yin settings were giving more erratic result than the default from descr, but that might depend on the audio file.  Do you have a few examples of varying toughness?
More soon!
Diemo

**Comment by christophertrapani** [comment]
Thanks, Diemo! I think it might be best to separate the problem into two
parts:

1.) getting the optimal algorithm — a single pipo process if possible — and
the most efficient way of writing the data into MuBu (is the line by line
scanning really the best option?)
2.) find the best default settings

for 1.), I realize there's a yin.periodicitythreshold in the first pipo,
but I think it might be helpful to be able to have a separate threshold for
computing markers and one for pitch estimation

2.) Good analysis values do vary a lot from one file to the next, and that
is a problem for analyzing folders with diverse files. I think one solution
— which I was hoping to implement myself — would be to have not a single
default set of values but something resembling the dropdown menu of 10-12
options that Audiosculpt uses: "low monophonic sound," "mid-register
polyphonic" for windowsize / analysis values — and maybe even some way of
scanning a sound file to guess which setting might be best?

I will try to put together a small folder of examples soon...

In larger terms: I am realizing that several pitch estimate-related
operations would be of interest: pitch estimate per grain, chord estimate
(have been getting cool results with chordino which I would love to share),
"first pitch" or "last pitch" of a longer grain... Just thinking out loud
to keep these ideas in mind...

Thanks!

**Comment by diemoschwarz** [comment]
Hi Chris, I got sidetracked by some missing handling of illegal and extreme values in pipo.chop.  
Only next week I'll be able to add a pipo that returns the median of a segment. That might already be effective when the input pitch is clean enough, which should be achievable by choosing the right yin threshold.  If that's not enough, as a next step, additional thresholds can be added.  I'd still be curious to pitch that against your test sounds...
Best!

**Comment by christophertrapani** [comment]
[pitchestimate-sound-examples.zip](https://github.com/ircam-ismm/catart-mubu/files/11925317/pitchestimate-sound-examples.zip)

Here's a collection of 4 sound examples (roughly 30 sec each, mix of mp3/wav, stereo/mono) similar to what I've been trying to segment.

A note about parameters: Even after many years, I have to admit I don't properly understand how to use filter size, odfmode, and onseg thresholds to optimize calculations. Would be interested in clarifying, both for my own understanding and to improve the documentation for others...

**Comment by diemoschwarz** [comment]
thanks for the files, Chris! the Greek one is the bane of segmenters...
I added a patch with spectral flux segmentation and median midi pitch all in one pipo graph, and some nice visualisation options.
You need these 2 updated pipo modules:
[Archive.zip](https://github.com/ircam-ismm/catart-mubu/files/11970025/Archive.zip)

**Reply by christophertrapani** [reply]
Unfortunately getting these errors: 

pipo.segment: could not load due to incorrect architecture
pipo.segmedian: could not load due to incorrect architecture

something else I need to do besides slipping them into the MuBu folder in the searchpath? thanks

**Comment by diemoschwarz** [comment]
Ah, so you're on M1/2.  Can you run Max under Rosetta until I get Riccardo to 
recompile them tomorrow?

**Reply by diemoschwarz** [reply]
Riccardo made you a prerelease here:
https://nubo.ircam.fr/index.php/s/SfkR3yMCb84KdfH

**Reply by christophertrapani** [reply]
Thanks — but this pre-release doesn't seem to have pipo.segmedian, right? I threw in the one you sent yesterday, but it's still only working via Rosetta

**Reply by diemoschwarz** [reply]
Riccardo will send it shortly.

**Comment by diemoschwarz** [comment]
While you wait, you can play with and optimise yin's thresholds interactively here:
[yin-explorer.maxpat.zip](https://github.com/ircam-ismm/catart-mubu/files/11981473/yin-explorer.maxpat.zip)

**Comment by christophertrapani** [comment]
Really cool and useful, Diemo — Thanks! 

Attached is the "classic" set of menu options for AudioSculpt settings, in XML format. I'm wondering if we can adapt these into a umenu that would give default attributes for a variety of sounds. Which values would you say are pertinent to recover? Willing to do the (pretty basic) programming work myself. Thanks again!

[AS-Settings.zip](https://github.com/ircam-ismm/catart-mubu/files/11981845/AS-Settings.zip)

**Reply by diemoschwarz** [reply]
as AS uses different algorithms, I'm afraid only the MinF0 could be relevant here, translated into winsize in ms to be independent of sr.

**Reply by diemoschwarz** [reply]
The maximum f0 searched for by yin is implicitly given by the downsampling (none: 22, 2x: 11, 4x: 5.5, 8x: 2.75 kHz) and could be cross-checked with the MaxF0 AS parameter, also for different SRs.

**Comment by RiccardoBorghesi** [comment]
Hi Christopher,
I added pipo.segmedian to the release.
You can download it from here:
https://nubo.ircam.fr/index.php/s/JTWY3B9ZwMcfB88

Best
Riccardo

**Reply by christophertrapani** [reply]
Grazie Riccardo, Vielen Danke Diemo! Let me make a new tutorial 3 next, with an updated, simpler module that now does a descr analysis based on the time column of the median track, then inserts the NoteNumber values as a constant there, so they can be used in mubu.concat along with other descr values

Thanks so much, and more soon — C

**Reply by diemoschwarz** [reply]
the example patch does the NN and median in one go. If the incoming yin values are as good as can be, I was hoping that would make the extra patch obsolete, maybe even without filtering. If the results are yet worse we can think how to introduce weighting/filtering into the median.

**Reply by christophertrapani** [reply]
Here's a new version with a far simplified intermediate object (camu.movepitchcolumn) that moves the pitch estimate data to descr as a constant column

Some thoughts:

1.) pitch estimate gives greatly improved results using markers derived from spectral differencing, but I wonder if it could be further filtered for more precise results. Also: with separate pipo process we could do marker detection with one window size and frequency analysis with one (larger) window size. Is this irrelevant now, or still something that could be incorporated?

Here's the media pitch per segment for the Easy-Turkey example. Accurate and useful, but I imagine with filtering we could see clusters of pitches better 
<img width="458" alt="Screenshot 2023-07-09 at 6 05 34 PM" src="https://github.com/ircam-ismm/catart-mubu/assets/35881355/6300a355-ff1d-4f9f-ad33-41e84a12e04f">

2.) the laborious process of scanning every line in a matrix and adding the const value in camu.movepitchcolumn: can it be streamlined at all? I have programmed it to work for only one buffer, 
but am unclear about the best strategy to make it scan each buffer in sequence.

Also wondering:

3.) How to name the constant
4.) How to add more than one constant column

[camu.movepitchcolumn.maxpat.zip](https://github.com/ircam-ismm/catart-mubu/files/11995588/camu.movepitchcolumn.maxpat.zip)
[catart-mubu-tutorial3-w-pitchestimate-mel-streamline-9July.maxpat.zip](https://github.com/ircam-ismm/catart-mubu/files/11995589/catart-mubu-tutorial3-w-pitchestimate-mel-streamline-9July.maxpat.zip)

**Reply by diemoschwarz** [reply]
Hi Chris, just a quick update, I've been on your use case all that time, trying to put everything in one pipo graph, but the multi-rate pipo parallel graph opens a Pandora's box of worms. It will be possible eventually, but for the moment I'm backtracking and separating segmentation from analysis again. I'll have a new patch shortly with amazing imubu displays thanks to Riccardo, who will send you a new release.

> 1.) pitch estimate further filtering

that would be doable in pipo.js, and then we could easily see how much improvements it brings and if it should be integrated into segmedian

> 2.) the laborious process of scanning every line in a matrix and adding the const value in camu.movepitchcolumn

This is 4 objects`getmatrix` in `mubu.track ... @matrixcol source` and `setmatrix 0` in `mubu.track ... @matrixcol target`, but it wont be necessary anymore

>  4.) How to add more than one constant column

pipo.conts accepts a list of column names

More soon, best
Diemo

**Reply by diemoschwarz** [reply]
Hi Chris, here's an updated version of mubu fixing many display things: https://nubo.ircam.fr/index.php/s/rrzWM9j9twccXcW
but for the latest patch you also need a new pipo.js [pipo.javascript.mxo.zip](https://github.com/ircam-ismm/catart-mubu/files/12259379/pipo.javascript.mxo.zip) that I can only build for Intel, so you'll have to use Rosetta until Riccardos is back from holidays.

The latest version of the patch [here](https://raw.githubusercontent.com/ircam-ismm/catart-mubu/wip-yinseg/patchers/tutorials/catart-mubu-tutorial3-w-pitchestimate-mel-separate.maxpat) is doing filtering of MIDI pitch by periodicity and loudness, before the median calculation (steps 3 and 4 in `camu.pitchestimate`). For step 5 _Filter out any pitches +/- 5 cents from median estimate_, I'd love to see an example where this actually improves the result, for the median should already do a close to optimal job of removing outliers.

Happy tryout! 
Diemo

**Reply by christophertrapani** [reply]
Diemo / Riccardo: Was a new M1-friendly build of mubu made, and could you send me a link to the latest if so? Just checking as I get back into this question after a long delay; thanks!

C
